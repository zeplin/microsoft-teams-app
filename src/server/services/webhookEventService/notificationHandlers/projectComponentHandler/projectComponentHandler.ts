import { NotificationHandler } from "../NotificationHandler";
import { MessageCard, commonTeamsCard } from "../teamsCardTemplates";
import {
    ProjectComponentCreateEvent,
    ProjectComponentVersionCreateEvent
} from "../../../../adapters/zeplin/types";
import { MEDIUM_DELAY } from "../constants";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../config";
import { getMacAppRedirectURL } from "../getMacAppRedirectURL";

const IMAGE_LIMIT = 5;

type Event = ProjectComponentCreateEvent | ProjectComponentVersionCreateEvent;

class ProjectComponentHandler extends NotificationHandler<Event> {
    delay = MEDIUM_DELAY;

    private getText(events: Event[]): string {
        const [{
            payload: {
                action,
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        name: componentName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added" : "updated";
        return events.length === 1
            ? `**${componentName}** is ${actionText} in _${projectName}_! 🏃‍♂️`
            : `**${events.length}${action === "created" ? " new" : ""} components** are ${actionText} in _${projectName}_! 🏃‍♂️`;
    }

    private getImages(events: Event[]): string[] {
        // Take last 5 screen images
        return events
            .sort((e1, e2) => e2.payload.timestamp - e1.payload.timestamp)
            .map(event => event.payload.resource.data.image?.original_url)
            .filter((val): val is string => Boolean(val))
            .slice(0, IMAGE_LIMIT);
    }

    private getWebappURL(events: Event[]): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/styleguide/components`;
        events.forEach(event => webappURL.searchParams.append("coid", event.payload.resource.id));
        return webappURL.toString();
    }

    private getMacAppURL(events: Event[]): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        return getMacAppRedirectURL(`${ZEPLIN_MAC_APP_URL_SCHEME}://components?pid=${projectId}&coids=${events.map(event => event.payload.resource.id).join(",")}`);
    }

    getTeamsMessage(events: Event[]): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            images: this.getImages(events),
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }

    shouldHandleEvent(event: Event): event is Event {
        return event.payload.action === "created" || event.payload.action === "version_created";
    }
}

export const projectComponentHandler = new ProjectComponentHandler();
