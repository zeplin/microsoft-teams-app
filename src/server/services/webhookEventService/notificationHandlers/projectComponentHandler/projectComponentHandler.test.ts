import { projectComponentHandler } from "./projectComponentHandler";
import {
    ProjectComponentCreateEvent,
    ProjectComponentVersionCreateEvent
} from "../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    componentId?: string;
    componentName?: string;
    imageUrl?: string;
    timestamp?: number;
}

function getDummyEvent({
    action = "created",
    componentId = "componentId",
    componentName = "componentName",
    imageUrl = "http://placehold.it/200",
    timestamp = 1
}: GetDummyEventParams = {}): ProjectComponentCreateEvent | ProjectComponentVersionCreateEvent {
    return {
        payload: {
            timestamp,
            action,
            context: {
                project: {
                    id: "projectId",
                    name: "projectName"
                }
            },
            resource: {
                id: componentId,
                data: {
                    id: componentId,
                    name: componentName,
                    image: {
                        thumbnails: {
                            small: imageUrl
                        }
                    }
                }
            }
        }
    } as ProjectComponentCreateEvent | ProjectComponentVersionCreateEvent;
}

describe("projectComponentHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        it("should images be sorted by timestamp and clamped to contain 5 non-empty urls", () => {
            const result = projectComponentHandler.getTeamsMessage([
                getDummyEvent({ timestamp: 6, imageUrl: "url6" }),
                getDummyEvent({ timestamp: 0, imageUrl: "" }),
                getDummyEvent({ timestamp: 5, imageUrl: "url5" }),
                getDummyEvent({ timestamp: 4, imageUrl: "url4" }),
                getDummyEvent({ timestamp: 3, imageUrl: "url3" }),
                getDummyEvent({ timestamp: 1, imageUrl: "url1" }),
                getDummyEvent({ timestamp: 2, imageUrl: "url2" })
            ]);
            const expectedImages = ["url6", "url5", "url4", "url3", "url2"];
            const { sections: [{ images }] } = result;
            expect(images?.map(image => image.image)).toEqual(expectedImages);
        });

        describe("for created event", () => {
            it("should match snapshot when there is only one event", () => {
                expect(projectComponentHandler.getTeamsMessage([
                    getDummyEvent()
                ])).toMatchSnapshot();
            });

            it("should match snapshot when there are more than one events", () => {
                expect(projectComponentHandler.getTeamsMessage([
                    getDummyEvent(),
                    getDummyEvent({
                        componentId: "componentId2"
                    })
                ])).toMatchSnapshot();
            });
        });

        describe("for version_created event", () => {
            it("should match snapshot when there is only one event", () => {
                expect(projectComponentHandler.getTeamsMessage([
                    getDummyEvent({
                        action: "version_created"
                    })
                ])).toMatchSnapshot();
            });

            it("should match snapshot when there are more than one events", () => {
                expect(projectComponentHandler.getTeamsMessage([
                    getDummyEvent({
                        action: "version_created"
                    }),
                    getDummyEvent({
                        action: "version_created",
                        componentId: "componentId2"
                    })
                ])).toMatchSnapshot();
            });
        });
    });
});
