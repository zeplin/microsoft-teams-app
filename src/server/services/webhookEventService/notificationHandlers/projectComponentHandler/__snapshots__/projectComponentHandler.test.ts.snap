// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectComponentHandler getTeamsMessage method for created event should match snapshot when there are more than one events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 new components** are added in _projectName_! ðŸ—ž",
  "sections": [
    {
      "images": [
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        },
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        }
      ]
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcomponents%3Fpid%3DprojectId%26coids%3DcomponentId%2CcomponentId2"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/components?coid=componentId&coid=componentId2"
        }
      ]
    }
  ]
}
`;

exports[`projectComponentHandler getTeamsMessage method for created event should match snapshot when there is only one event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**componentName** is added in _projectName_! ðŸ—ž",
  "sections": [
    {
      "images": [
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        }
      ]
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcomponents%3Fpid%3DprojectId%26coids%3DcomponentId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/components?coid=componentId"
        }
      ]
    }
  ]
}
`;

exports[`projectComponentHandler getTeamsMessage method for version_created event should match snapshot when there are more than one events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 components** are updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "images": [
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        },
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        }
      ]
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcomponents%3Fpid%3DprojectId%26coids%3DcomponentId%2CcomponentId2"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/components?coid=componentId&coid=componentId2"
        }
      ]
    }
  ]
}
`;

exports[`projectComponentHandler getTeamsMessage method for version_created event should match snapshot when there is only one event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**componentName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "images": [
        {
          "image": "http://placehold.it/200",
          "title": "Section Image"
        }
      ]
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcomponents%3Fpid%3DprojectId%26coids%3DcomponentId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/components?coid=componentId"
        }
      ]
    }
  ]
}
`;
